BUG: pivot_table with overlapping values

it176131 opened the pull request:
- [x] closes #57876
- [x] closes #61292
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

it176131 committed:
modified:   pandas/tests/reshape/test_pivot_multilevel.py
	- Added two tests, :func:`test_pivot_table_values_in_columns` and :func:`test_pivot_table_values_in_index`, to ensure that the `values` param is still used when the argument is shared between the `columns` and `values` params, and `index` and `values` params.

it176131 committed:
modified:   pandas/core/reshape/pivot.py
	- Added condition to :func:`__internal_pivot_table` to aggregate `values` explicitly if `values` were passed, otherwise aggregate all remaining columns. This allows the tests :func:`test_pivot_table_values_in_columns` and :func:`test_pivot_table_values_in_index` in test_pivot_multilevel.py to pass.

it176131 committed:
modified:   pandas/tests/reshape/test_pivot.py
	- Added test :func:`test_pivot_table_values_as_two_params` to test that the updates in pivot.py result in expected results, satisfying GH issue #57876.

it176131 committed:
modified:   pandas/tests/reshape/test_pivot.py
	- Added GH issue comment to test :func:`test_pivot_table_values_as_two_params`.

it176131 committed:
modified:   pandas/tests/reshape/test_pivot_multilevel.py
	- Combined tests :func:`test_pivot_table_values_in_columns` and :func:`test_pivot_table_values_in_index` into a single parametrized test, :func:`test_pivot_table_multiindex_values_as_two_params` to reduce duplicate setup code.

it176131 committed:
modified:   pandas/tests/reshape/test_pivot_multilevel.py
	- Added GH issue #61292 as comment to test :func:`test_pivot_table_multiindex_values_as_two_params`.

it176131 committed:
modified:   pandas/core/reshape/pivot.py
	- Simplified proposed logic in :func:`__internal_pivot_table`.

it176131 committed:
modified:   pandas/core/reshape/pivot.py
	- Added GH issue numbers to new logic in :func:`__internal_pivot_table`.

it176131 commented:
Should I update the `doc/source/whatsnew/v2.3.0.rst` or `doc/source/whatsnew/v3.0.0.rst` file?

it176131 committed:
modified:   pandas/core/reshape/pivot.py
	- Added ignore-comment to silence mypy error in :func:`__internal_pivot_table`.
	- Added TODO-comment stating that the :meth:`DataFrameGroupBy.__getitem__` should be overloaded to match the pandas-stubs type declarations, informing mypy that the type is correct given `values` is a list.

rhshadrach commented:
@it176131 - update v3.0.0. Also due to an erroneous commit with a large file, we've had to force-push to the main branch recently. You'll need to fix this up by removing the erroneous commit from your history as well. Check out the diff for more.

@datapythonista - if we're going to force-push to a contributors branch, I think a comment detailing what's going on would be helpful.

it176131 commented:
@rhshadrach I believe I've reverted the changes as requested. Please lmk if I did it incorrectly.

it176131 committed:
modified:   doc/source/whatsnew/v3.0.0.rst
	- Added pivot_table bug to Bugs/Reshaping section referencing issues #57876 and #61292.

datapythonista commented:
Oh, so GitHub added the image we removed to the open PRs? That explains why it was in that PR...

it176131 commented:
@rhshadrach [this check](https://github.com/pandas-dev/pandas/actions/runs/14518139891/job/40732136143?pr=61293) appears to have failed due to a timeout—is there a way to restart it?

rhshadrach commented on pandas/core/reshape/pivot.py:
```diff
@@ -336,6 +336,16 @@ def __internal_pivot_table(
     grouped = data.groupby(keys, observed=observed, sort=sort, dropna=dropna)
+    if values_passed:
+        # GH#57876 and GH#61292
+        # Explicitly aggregate ``values``.
```
I think this comment can be removed (just L341; not L340).

rhshadrach commented on pandas/tests/reshape/test_pivot.py:
```diff
@@ -2554,6 +2554,39 @@ def test_pivot_table_index_and_column_keys_with_nan(self, dropna):
         tm.assert_frame_equal(left=result, right=expected)
 
+    @pytest.mark.parametrize(
+        argnames=["index", "columns"],
+        argvalues=[("Category", "Value"), ("Value", "Category")],
```
Just to agree with the style of the rest of the test-suite, can you do this instead:

```suggestion
        "index, columns",
        [("Category", "Value"), ("Value", "Category")],
```

rhshadrach commented on pandas/tests/reshape/test_pivot.py:
```diff
@@ -2554,6 +2554,39 @@ def test_pivot_table_index_and_column_keys_with_nan(self, dropna):
+            ]
+            expected = DataFrame(data=e_data, index=val_index, columns=cat_index)
+
+        tm.assert_frame_equal(left=result, right=expected)
```
Similar comment on style - I see this is in the test above, but that is the only other test that does this.
```suggestion
        tm.assert_frame_equal(result, expected)
```

it176131 replied:
I actually wrote that other test a couple weeks ago 😅

rhshadrach commented on pandas/tests/reshape/test_pivot_multilevel.py:
```diff
@@ -250,3 +250,46 @@ def test_pivot_df_multiindex_index_none():
+
+
+@pytest.mark.parametrize(
+    argnames=["index", "columns"],
+    argvalues=[("index", ["col", "value"]), (["index", "value"], "col")],
```
Same as above.

rhshadrach commented on pandas/tests/reshape/test_pivot_multilevel.py:
```diff
@@ -250,3 +250,46 @@ def test_pivot_df_multiindex_index_none():
+        e_cols = MultiIndex.from_arrays(
+            arrays=[[1, 1, 2, 2], [50, 100, 100, 200]], names=["col", "value"]
+        )
+        expected = pd.DataFrame(data=e_data, index=e_index, columns=e_cols)
```
Can remove this line if you also dedent L293.

it176131 replied:
Good catch 👍

rhshadrach commented on pandas/tests/reshape/test_pivot_multilevel.py:
```diff
@@ -250,3 +250,46 @@ def test_pivot_df_multiindex_index_none():
+        e_cols = Index(data=[1, 2], name="col")
+        expected = pd.DataFrame(data=e_data, index=e_index, columns=e_cols)
+
+    tm.assert_frame_equal(left=result, right=expected)
```
Same as above.

rhshadrach reviewed (changes requested):
Looking good! Mostly nit-picks below. I'd ignore the free-threading workflow for now.

it176131 committed:
modified:   pandas/core/reshape/pivot.py
	- Moved and simplified mypy comment per feedback.

it176131 committed:
modified:   pandas/core/reshape/pivot.py
	- Removed comment about explicit aggregation per feedback.

it176131 committed:
modified:   pandas/tests/reshape/test_pivot.py
	- Removed param names and updated `argnames` arg per feedback in parametrized marker.

it176131 committed:
modified:   pandas/tests/reshape/test_pivot.py
	- Removed param names in favor of implicit args per feedback.

it176131 committed:
modified:   pandas/tests/reshape/test_pivot_multilevel.py
	- Removed param names and updated arg for `argnames` in parametrized marker per feedback.

it176131 committed:
modified:   pandas/tests/reshape/test_pivot_multilevel.py
	- Reduced `expected` assignments from two to one per feedback.

it176131 committed:
modified:   pandas/tests/reshape/test_pivot_multilevel.py
	- Removed param names in favor of implicit args per feedback.

rhshadrach reviewed (approved):
lgtm

mroeschke commented on pandas/tests/reshape/test_pivot.py:
```diff
@@ -2554,6 +2554,39 @@ def test_pivot_table_index_and_column_keys_with_nan(self, dropna):
+        nan = np.nan
+        cat_index = Index(data=["A", "B"], name="Category")
+        val_index = Index(data=[10, 20, 40, 50], name="Value")
+        if request.node.callspec.id == "values-and-columns":
+            e_data = [
+                [1.0, nan, 1.0, nan],
+                [nan, 1.0, nan, 1.0],
+            ]
```
Can you put `e_data` in the `pytest.mark.parametrize` instead of checking `request.node.callspec.id` (same as below)?

it176131 committed:
Merge branch 'main' into bug-pivot-table

it176131 committed:
modified:   pandas/tests/reshape/test_pivot_multilevel.py
	- Moved e_data, e_index, and e_cols to parametrized marker instead of declaring inside the test :func:`test_pivot_table_multiindex_values_as_two_params`.

it176131 committed:
modified:   pandas/tests/reshape/test_pivot.py
	- Moved `expected` setup to parametrized marker instead of in the test :meth:`TestPivotTable.test_pivot_table_values_as_two_params`.

it176131 committed:
modified:   pandas/tests/reshape/test_pivot.py
	- Removed walrus operator declarations in parametrized marker for test :meth:TestPivotTable.test_pivot_table_values_as_two_params`. Appears related to this mypy issue -> https://github.com/python/mypy/issues/17377.

it176131 committed:
modified:   pandas/tests/reshape/test_pivot_multilevel.py
	- Removed walrus operator declarations as I'm sure mypy would raise an issue with it given that it did in test_pivot.py (see commit 52cf56011e).

mroeschke commented:
Thanks @it176131
